### PR TITLE
magento/magento2#28481: GraphQl. updateCustomer allows to set any INT value in gender argument

### LINKED ISSUE #28481: GraphQl. updateCustomer allows to set any INT value in gender argument
### Preconditions (*)

1. Magento with a sample data

### Steps to reproduce (*)

1. Generate a customer token

```
mutation {
  generateCustomerToken(
    email: "roni_cost@example.com",
    password: "roni_cost3@example.com"
  ) {
    token
  }
}
```

2. Set a `gender` value for a customer

```
mutation {
  updateCustomer(
    input: {
      gender: 13
    }
  ) {
    customer {
      firstname
      gender
    }
  }
}
```

### Expected result (*)

1. Customer `gender` attribute may have one of the following values:

- Male (1)
- Female (2)
- Not specified (3)

### Actual result (*)

```
{
  "data": {
    "updateCustomer": {
      "customer": {
        "firstname": "Veronica",
        "gender": 13
      }
    }
  }
}
```

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by atwixfirster**
### Description (*)

PR fixes [GraphQl. updateCustomer allows to set any INT value in gender argument](https://github.com/magento/magento2/issues/28481)

### Fixed Issues (if relevant)

1. Fixes magento/magento#228481: GraphQl. updateCustomer allows to set any INT value in gender argument

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by atwixfirster** [comment]
@magento run all tests

**Comment by atwixfirster** [comment]
@magento run Database Compare

**Comment by atwixfirster** [comment]
@magento run Static Tests,Unit Tests,WebAPI Tests,Integration Tests

**Comment by atwixfirster** [comment]
@magento run Static Tests,Unit Tests,WebAPI Tests,Integration Tests

**Comment by atwixfirster** [comment]
@magento run Static Tests

**Comment by dmytro-ch** [comment]
@magento run all tests

**Review by dmytro-ch — CHANGES_REQUESTED** [review]
Hello @atwixfirster, could you please update your working branch and resolve merge conflicts?

**Comment by atwixfirster** [comment]
@magento run all tests

File: app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php [hunk 1]
```diff
@@ -32,25 +41,40 @@ class ValidateCustomerData
      *
      * @param GetAllowedCustomerAttributes $getAllowedCustomerAttributes
      * @param EmailAddressValidator $emailAddressValidator
+     * @param CustomerMetadataInterface $customerMetadata
      */
     public function __construct(
         GetAllowedCustomerAttributes $getAllowedCustomerAttributes,
-        EmailAddressValidator $emailAddressValidator
+        EmailAddressValidator $emailAddressValidator,
+        CustomerMetadataInterface $customerMetadata
```
Comment (prabhuram93) [inline app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php]: Please make sure to validate constructor params for null, when adding a new param to an existing class. something in the lines of  $customerMetadata ?: \Magento\Framework\App\ObjectManager::getInstance()
            ->get(CustomerMetadataInterface::class);

File: app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php [hunk 2]
```diff
@@ -32,25 +41,40 @@ class ValidateCustomerData
...
         $this->emailAddressValidator = $emailAddressValidator;
+        $this->customerMetadata = $customerMetadata;
     }
 
     /**
      * Validate customer data
      *
      * @param array $customerData
-     *
-     * @return void
```
Comment (prabhuram93) [inline app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php:49]: still returns void

File: app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php [hunk 3]
```diff
@@ -32,25 +41,40 @@ class ValidateCustomerData
...
-     *
      * @throws GraphQlInputException
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
      */
     public function execute(array $customerData): void
+    {
+        $this->validateRequiredArguments($customerData);
```
Comment (prabhuram93) [inline app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php]: I like the idea here. Would be nice to have an interface that has a validate method to accept array $customerData and validator classes which implement the interface corresponding to arguments, email and gender. And you can pass in the validators via di.xml as arguments.

**Review by prabhuram93 — CHANGES_REQUESTED** [review]

File: app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php (hunk 2 above)
Comment (atwixfirster) [reply app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php:49]: **removed**

File: app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php (hunk 3 above)
Comment (atwixfirster) [reply app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php]: @prabhuram93 
Could you please verify a new solution with defined validators in the `di.xml`?

Thank you!

**Comment by atwixfirster** [comment]
@magento run all tests

File: app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php (hunk 3 above)
Comment (prabhuram93) [reply app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php]: Cool. looks good!

File: app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php (hunk 1 above)
Comment (prabhuram93) [reply app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php]: This comment is outdated now.

**Review by prabhuram93 — APPROVED** [review]

**Comment by atwixfirster** [comment]
@magento run all tests

**Comment by atwixfirster** [comment]
@magento Static Tests

File: app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php [hunk 4]
```diff
@@ -15,6 +18,11 @@
  */
 class ValidateCustomerData
 {
+    /**
+     * @var ValidateCustomerDataInterface[]
+     */
+    protected $validators = [];
```
Comment (dthampy) [inline app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php]: Please change the variable to private.

**Comment by dthampy** [comment]
@atwixfirster,  Magento Health Index build was failing. I left a comment in the appropriate file. If you could take care of that and run all the magento tests, I think we should be good. We can proceed with Merge.

**Comment by avattam06** [comment]
@magento run all tests

File: app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php (hunk 4 above)
Comment (atwixfirster) [reply app/code/Magento/CustomerGraphQl/Model/Customer/ValidateCustomerData.php]: Thank you for the explanations, @dthampy !

**done**

**Comment by atwixfirster** [comment]
@magento run all tests

**Review by dthampy — APPROVED** [review]

**Pull request merged**